blockchain/sync_manager/src/sync.rs

```diff
@@ -122,6 +146,142 @@ impl<'a> Syncer<'a> {
+                    // sequence equality check
+                    if *sequence != msg.sequence() {
+                        return Err(Error::Validation("Sequences are not equal".to_string()));
+                    }
+
+                    // sufficient funds check
+                    if balance < &msg.required_funds() {
+                        return Err(Error::Validation(
+                            "Insufficient funds for message execution".to_string(),
+                        ));
+                    }
+                    // update balance and increment sequence by 1
+                    MsgMetaData {
+                        balance: balance - msg.required_funds(),
+                        sequence: sequence + 1,
+                    }
+                }
+                // MsgMetaData not found with provided address key, insert sequence and balance with address as key
+                _ => {
+                    let actor = tree.get_actor(msg.from());
+                    if let Some(act) = actor {
+                        MsgMetaData {
+                            sequence: *act.sequence(),
+                            balance: act.balance().clone(),
+                        }
+                    } else {
+                        return Err(Error::State(
+                            "Could not retrieve actor from state tree".to_string(),
+                        ));
+                    }
+                }
+            };
+            // update hash map with updated state
+            msg_meta_data
+                .metadata
+                .insert(msg.from().clone(), updated_state);
+            Ok(())
+        }
+
+        let mut msg_meta_data = MsgCheck::new();
+        // TODO retrieve tipset state and load state tree
+        // temporary
+        let tree = HamtStateTree::default();
+        // loop through bls messages and check msg validity
+        for m in block.bls_msgs() {
+            check_msg(m.clone(), &mut msg_meta_data, &tree)?;
+        }
+        // loop through secp messages and check msg validity and signature
+        for m in block.secp_msgs() {
+            check_msg(m.clone(), &mut msg_meta_data, &tree)?;
+            // signature validation
+            if !is_valid_signature(&m.cid()?.to_bytes(), m.from(), m.signature()) {
+                return Err(Error::Validation(
+                    "Message signature is not valid".to_string(),
+                ));
+            }
+        }
+        // validate message root from header matches message root
+        let sm_root = self.compute_msg_data(&block)?;
+        if block.to_header().messages() != &sm_root {
+            return Err(Error::InvalidRoots);
+        }
+
+        Ok(())
+    }
+
+    /// Validates block semantically according to https://github.com/filecoin-project/specs/blob/6ab401c0b92efb6420c6e198ec387cf56dc86057/validation.md
+    pub fn validate(&self, block: Block) -> Result<(), Error> {
```

Comment:
What's the rationale for this being a method on Syncer? At first glance is feels like this logic should be associated with `Block`

Reply:
A few reasons, avoiding circular dependencies as the validation requires a lot of pieces from other crates (especially when we complete all checks) of which those crates are used already through the sync crate, the `syncer.rs` file provides the functionality required for syncing which validation is a part of whereas, block.rs (at least right now) just defines the Block structures, the Block of which we are validating will be passed through the syncer via `inform_new_head`, and it's also how Lotus has it in place.

Reply:
As discussed, given our current crate design we run into circular dependencies related to the `chain_store` if this function is moved to `block.rs`. If we want to change how our crates are organized to allow for the movement of this function including all the remaining validation checks (e.g. state which all requires access to chain_store) I think that should be put forth in a separate PR.

Reply:
I am fine with this now as well because it relies on the syncer to retrieve the full tipset, so would in fact introduce a circular dep for blocks to use the sync_manager crate (I didn't notice this when we chatted this morning)